java-checks/src/main/java/org/sonar/java/checks/xml/struts/ActionNumberCheck.java

```diff
@@ -41,30 +38,25 @@
     defaultValue = "" + DEFAULT_MAXIMUM_NUMBER_FORWARDS)
   public int maximumForwards = DEFAULT_MAXIMUM_NUMBER_FORWARDS;
 
-  private XPathExpression actionsExpression;
-  private XPathExpression forwardsFromActionExpression;
+  private XPathExpression actionsExpression = getXPathExpression("struts-config/action-mappings/action");
+  private XPathExpression forwardsFromActionExpression = getXPathExpression("forward");
 
   @Override
-  public void precompileXPathExpressions(XmlCheckContext context) {
-    actionsExpression = context.compile("struts-config/action-mappings/action");
-    forwardsFromActionExpression = context.compile("forward");
-  }
-
-  @Override
-  public void scanFileWithXPathExpressions(XmlCheckContext context) {
-    for (Node action : context.evaluateOnDocument(actionsExpression)) {
-      Iterable<Node> extraForwards = Iterables.skip(context.evaluate(forwardsFromActionExpression, action), maximumForwards);
-      if (!Iterables.isEmpty(extraForwards)) {
-        List<XmlCheckContext.XmlDocumentLocation> secondaries = new LinkedList<>();
-        for (Node forward : extraForwards) {
-          secondaries.add(new XmlCheckContext.XmlDocumentLocation("Move this forward to another action.", forward));
+  protected void scanFile(XmlFile xmlFile) {
+    evaluateAsList(actionsExpression, xmlFile.getNamespaceUnawareDocument())
+      .forEach(node -> {
```

Comment:
I'm not a big fan of nested lambdas. Can you extract this part into a dedicated method?
It will then be simply something such as:
```
 evaluateAsList(actionsExpression, xmlFile.getNamespaceUnawareDocument()).forEach(this::checkAction);
```